### PR TITLE
added primary id validation

### PR DESCRIPTION
**Pull request by ammokhov**
*Issue #, if available:* #361 

*Description of changes:*

we do not allow mutable updates on primary identifier, instead it could be updated immutably, which causes the replacement.
However, primary identifier could be specified in two ways: 
* `readOnly` - auto generated by handler or api
* `createOnly` - specified by the user and cant be further mutably updated

Since, we would want to enforce the schema modeling it makes sense to add this to schema validation to provide definition correctness. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rjlohan** [comment]
Should this validation be added to https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/tree/master/src as well/instead?

**Comment by jotompki** [comment]
> Should this validation be added to https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/tree/master/src as well/instead?

Agree with this we should just schematize this

File: src/rpdk/core/data_loaders.py [hunk 1]
```diff
@@ -106,6 +113,16 @@ def load_resource_spec(resource_spec_file):  # noqa: C901
         LOG.debug("Resource spec validation failed", exc_info=True)
         raise SpecValidationError(str(e)) from e
 
+    in_readonly = _is_in(resource_spec, "readOnlyProperties")
+    in_createonly = _is_in(resource_spec, "createOnlyProperties")
+
+    primary_id = resource_spec["primaryIdentifier"]
+    if not in_readonly(primary_id) and not in_createonly(primary_id):
```
Comment (jotompki) [inline src/rpdk/core/data_loaders.py:120]: can simplify a bit to make condition `primaryIdentifier - (createOnlyProperties | readOnlyProperties)`  where they are all sets

**Comment by ammokhov** [comment]
> 
> Should this validation be added to https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/tree/master/src as well/instead? 

> Agree with this we should just schematize this

as discussed offline, we cant really schematize the condition, so this change should go in both repos. 
However, if we start raising this exception it might break the registration, so alternatively we can log a warning to highlight the issue

File: src/rpdk/core/data_loaders.py (hunk 1 above)
Comment (ammokhov) [reply src/rpdk/core/data_loaders.py:120]: used closure to reuse the subset operation but we might use a disjoint of `createOnly | readOnly`

**Comment by ammokhov** [comment]
> why do we want to change this to warning?

we want to inform a customer that id should be either createonly or readonly but if we start throwing an exception it will block existing resources from registration. warning still informs but make it backward compatible

File: src/rpdk/core/data_loaders.py (hunk 1 above)
Comment (PatMyron) [reply src/rpdk/core/data_loaders.py:120]: [not sure this is handling composite primaryIdentifiers correctly](https://sage.amazon.com/questions/1025861)

File: tests/data/schema/valid/invalid_primary_id.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+    "typeName": "AWS::Valid::TypeName",
```
Comment (PatMyron) [inline tests/data/schema/valid/invalid_primary_id.json:2]: Was this schema to test https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/issues/98?
`typeName` should be renamed and should switch directories if so

**Pull request merged**